### PR TITLE
Workflow consolidation for GitHub actions

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74406
* __->__ #74306

This PR implements the proposal in https://fb.quip.com/RcLZAOb9k0qI,
which calls for using GitHub's "normal" means of composition (composite
actions, reusable workflows) instead of codegen.

### CODE REVIEW

File: .github/workflows/docker-builds.yml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  AWS_DEFAULT_REGION: us-east-1
+
+jobs:
+  docker-build:
+    runs-on: [self-hosted, linux.2xlarge]
+    timeout-minutes: 240
+    strategy:
+      matrix:
+        include:
+          - docker-image-name: pytorch-linux-bionic-cuda10.2-cudnn7-py3.9-gcc7
```
[the remark this answers is not in the document]
Comment (suo) [reply .github/workflows/docker-builds.yml:27]: the name list? Yeah it's not ideal. Although to be clear, it's hard-coded today as well in generate_ci_workflows.py, just indirectly :P

[the remark this answers is not in the document]
Comment (seemethere) [reply .github/workflows/docker-builds.yml:27]: Somewhat, but the good thing about `generate_ci_workflows.py` is that you didn't have to think about it when adding a new docker image, it just automatically added it to the list when you declared a new job

[the remark this answers is not in the document]
Comment (suo) [reply .github/workflows/docker-builds.yml:27]: Hmmm that makes sense. Is there a canonical list anywhere of all the docker images? If we want to make it more maintainable, one thing we can do is generate this matrix from a file (using a technique similar to what we used to do for `generate-test-matrix`). That way we don't hard-code it here but retrieve the list of images to build from some other source of truth.

**Review by janeyx99 — APPROVED** [review]

**Pull request closed without merging**